### PR TITLE
Deprecate Template Mgt API

### PR DESCRIPTION
**Pull request by GANGANI**
Related issue: https://github.com/wso2/product-is/issues/8621

### CODE REVIEW

File: components/template-mgt/org.wso2.carbon.identity.template.mgt/src/main/java/org/wso2/carbon/identity/template/mgt/TemplateManager.java [hunk 1]
```diff
@@ -30,6 +30,7 @@
  *
  * @since 1.0.0
  */
+@Deprecated
```
Comment (emswbandara) [inline components/template-mgt/org.wso2.carbon.identity.template.mgt/src/main/java/org/wso2/carbon/identity/template/mgt/TemplateManager.java:33]: do not deprecate template mgt core classes. only deprecate template mgt endpoint.

Comment (GANGANI) [reply components/template-mgt/org.wso2.carbon.identity.template.mgt/src/main/java/org/wso2/carbon/identity/template/mgt/TemplateManager.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/2996/commits/51d084a813c79c52636b52d39f06a834dcb09ed1

File: components/template-mgt/org.wso2.carbon.identity.template.mgt.endpoint/src/main/java/org/wso2/carbon/identity/template/mgt/endpoint/impl/TemplatesApiServiceImpl.java [hunk 2]
```diff
@@ -47,98 +47,39 @@
...
+@Deprecated
+public class TemplatesApiServiceImpl extends TemplatesApiService {
 
     private static final Log LOG = LogFactory.getLog(TemplatesApiServiceImpl.class);
 
     @Override
     public Response addTemplate(TemplateDTO template) {
 
-        try {
-            TemplateResponseDTO response = postTemplate(template);
```
Comment (emswbandara) [inline components/template-mgt/org.wso2.carbon.identity.template.mgt.endpoint/src/main/java/org/wso2/carbon/identity/template/mgt/endpoint/impl/TemplatesApiServiceImpl.java:59]: shall we keep the old code as it is without removing? we can return forbidden from the beginning of the function

File: components/template-mgt/org.wso2.carbon.identity.template.mgt/src/test/java/org/wso2/carbon/identity/template/mgt/TemplateManagerImplTest.java [hunk 3]
```diff
@@ -53,6 +53,7 @@
 import static org.wso2.carbon.identity.template.mgt.util.TestUtils.mockDataSource;
 import static org.wso2.carbon.identity.template.mgt.util.TestUtils.spyConnection;
 
+@Deprecated
```
Comment (emswbandara) [inline components/template-mgt/org.wso2.carbon.identity.template.mgt/src/test/java/org/wso2/carbon/identity/template/mgt/TemplateManagerImplTest.java:56]: this is also a test class of a core class.. let's keep it

File: components/template-mgt/org.wso2.carbon.identity.template.mgt/src/test/java/org/wso2/carbon/identity/template/mgt/dao/impl/TemplateManagerDAOImplTest.java [hunk 4]
```diff
@@ -47,6 +47,7 @@
 import static org.wso2.carbon.identity.template.mgt.util.TestUtils.mockDataSource;
 import static org.wso2.carbon.identity.template.mgt.util.TestUtils.spyConnection;
 
+@Deprecated
```
Comment (emswbandara) [inline components/template-mgt/org.wso2.carbon.identity.template.mgt/src/test/java/org/wso2/carbon/identity/template/mgt/dao/impl/TemplateManagerDAOImplTest.java:50]: remove deprecated tag from the classes in org.wso2.carbon.identity.template.mgt module

File: components/template-mgt/org.wso2.carbon.identity.template.mgt.endpoint/src/main/java/org/wso2/carbon/identity/template/mgt/endpoint/impl/TemplatesApiServiceImpl.java (hunk 2 above)
Comment (GANGANI) [reply components/template-mgt/org.wso2.carbon.identity.template.mgt.endpoint/src/main/java/org/wso2/carbon/identity/template/mgt/endpoint/impl/TemplatesApiServiceImpl.java:59]: can't keep the old code as it shows error "unreachable statement" when returning forbidden from the beginning of the function.

File: components/template-mgt/org.wso2.carbon.identity.template.mgt/src/test/java/org/wso2/carbon/identity/template/mgt/TemplateManagerImplTest.java (hunk 3 above)
Comment (GANGANI) [reply components/template-mgt/org.wso2.carbon.identity.template.mgt/src/test/java/org/wso2/carbon/identity/template/mgt/TemplateManagerImplTest.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/2996/commits/1aa997d374f2254338f5287b2bc2fef815104ff3

File: components/template-mgt/org.wso2.carbon.identity.template.mgt/src/test/java/org/wso2/carbon/identity/template/mgt/dao/impl/TemplateManagerDAOImplTest.java (hunk 4 above)
Comment (GANGANI) [reply components/template-mgt/org.wso2.carbon.identity.template.mgt/src/test/java/org/wso2/carbon/identity/template/mgt/dao/impl/TemplateManagerDAOImplTest.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/2996/commits/1aa997d374f2254338f5287b2bc2fef815104ff3

File: components/template-mgt/org.wso2.carbon.identity.template.mgt/src/test/java/org/wso2/carbon/identity/template/mgt/model/TemplateInfoTest.java [hunk 5]
```diff
@@ -21,6 +21,7 @@
 import org.testng.Assert;
 import org.testng.annotations.Test;
 
+
```
Comment (emswbandara) [inline components/template-mgt/org.wso2.carbon.identity.template.mgt/src/test/java/org/wso2/carbon/identity/template/mgt/model/TemplateInfoTest.java:24]: remove unncessary lines

File: components/template-mgt/org.wso2.carbon.identity.template.mgt/src/test/java/org/wso2/carbon/identity/template/mgt/model/TemplateTest.java [hunk 6]
```diff
@@ -22,6 +22,7 @@
 import org.testng.annotations.Test;
 import org.wso2.carbon.identity.template.mgt.TemplateMgtConstants;
 
+
```
Comment (emswbandara) [inline components/template-mgt/org.wso2.carbon.identity.template.mgt/src/test/java/org/wso2/carbon/identity/template/mgt/model/TemplateTest.java:25]: remove unncessary lines

File: components/template-mgt/org.wso2.carbon.identity.template.mgt/src/main/java/org/wso2/carbon/identity/template/mgt/TemplateMgtConstants.java [hunk 7]
```diff
@@ -43,6 +43,9 @@
     public static final String HEADER_CONTENT_TYPE = "Content-Type";
     public static final String DEFAULT_RESPONSE_CONTENT_TYPE = APPLICATION_JSON;
     public static final String STATUS_BAD_REQUEST_MESSAGE_DEFAULT = "Bad Request";
+    public static final String FORBIDDEN = "Forbidden";
+    public static final String FORBIDDEN_ERROR_CODE = "403";
+    public static final String FORBIDDEN_ERROR_MESSAGE= "Template-mgt API is not supported";
```
Comment (emswbandara) [inline components/template-mgt/org.wso2.carbon.identity.template.mgt/src/main/java/org/wso2/carbon/identity/template/mgt/TemplateMgtConstants.java:48]: shall we say Template-mgt API is restricted or something similar

File: components/template-mgt/org.wso2.carbon.identity.template.mgt/src/test/java/org/wso2/carbon/identity/template/mgt/model/TemplateInfoTest.java (hunk 5 above)
Comment (GANGANI) [reply components/template-mgt/org.wso2.carbon.identity.template.mgt/src/test/java/org/wso2/carbon/identity/template/mgt/model/TemplateInfoTest.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/2996/commits/59a224166a3a049623549bb53ac550322a2eaee0

File: components/template-mgt/org.wso2.carbon.identity.template.mgt/src/main/java/org/wso2/carbon/identity/template/mgt/TemplateMgtConstants.java (hunk 7 above)
Comment (GANGANI) [reply components/template-mgt/org.wso2.carbon.identity.template.mgt/src/main/java/org/wso2/carbon/identity/template/mgt/TemplateMgtConstants.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/2996/commits/59a224166a3a049623549bb53ac550322a2eaee0

File: components/template-mgt/org.wso2.carbon.identity.template.mgt/src/test/java/org/wso2/carbon/identity/template/mgt/model/TemplateTest.java (hunk 6 above)
Comment (GANGANI) [reply components/template-mgt/org.wso2.carbon.identity.template.mgt/src/test/java/org/wso2/carbon/identity/template/mgt/model/TemplateTest.java]: Fixed 59a2241

File: components/template-mgt/org.wso2.carbon.identity.template.mgt.endpoint/src/main/java/org/wso2/carbon/identity/template/mgt/endpoint/impl/TemplatesApiServiceImpl.java [hunk 8]
```diff
@@ -213,6 +154,10 @@ private Response handleBadRequestResponse(TemplateManagementClientException e) {
         throw TemplateEndpointUtils.buildBadRequestException(e.getMessage(), e.getErrorCode(), LOG, e);
     }
 
+    private Response handleForbiddenRequest() {
```
Comment (emswbandara) [inline components/template-mgt/org.wso2.carbon.identity.template.mgt.endpoint/src/main/java/org/wso2/carbon/identity/template/mgt/endpoint/impl/TemplatesApiServiceImpl.java:157]: add new line after method declaration

Comment (GANGANI) [reply components/template-mgt/org.wso2.carbon.identity.template.mgt.endpoint/src/main/java/org/wso2/carbon/identity/template/mgt/endpoint/impl/TemplatesApiServiceImpl.java:157]: Fixed https://github.com/wso2/carbon-identity-framework/pull/2996/commits/a88ebda2efe398dc4a2c32556f7c6a3a3d7ea958

File: components/template-mgt/org.wso2.carbon.identity.template.mgt.endpoint/src/main/java/org/wso2/carbon/identity/template/mgt/endpoint/impl/TemplatesApiServiceImpl.java [hunk 9]
```diff
@@ -47,98 +47,39 @@
 import static org.wso2.carbon.identity.template.mgt.TemplateMgtConstants.ErrorMessages.ERROR_CODE_UNEXPECTED;
 import static org.wso2.carbon.identity.template.mgt.TemplateMgtConstants.ErrorMessages.ERROR_CODE_USER_NOT_AUTHORIZED;
 
-public class
-TemplatesApiServiceImpl extends TemplatesApiService {
+@Deprecated
+public class TemplatesApiServiceImpl extends TemplatesApiService {
```
Comment (tharindu-bandara) [inline components/template-mgt/org.wso2.carbon.identity.template.mgt.endpoint/src/main/java/org/wso2/carbon/identity/template/mgt/endpoint/impl/TemplatesApiServiceImpl.java:51]: @GANGANI Any reason to change the access modifier for the method? We only want to deprecate the class right?

File: components/template-mgt/org.wso2.carbon.identity.template.mgt.endpoint/src/main/java/org/wso2/carbon/identity/template/mgt/endpoint/impl/TemplatesApiServiceImpl.java [hunk 10]
```diff
@@ -47,98 +47,39 @@
...
-                    .entity(response)
-                    .build();
-        } catch (TemplateManagementClientException e) {
-            return handleBadRequestResponse(e);
-        } catch (TemplateManagementException e) {
-            return handleServerErrorResponse(e);
-        } catch (Throwable throwable) {
-            return handleUnexpectedServerError(throwable);
-        }
+        return handleForbiddenRequest();
```
Comment (tharindu-bandara) [inline components/template-mgt/org.wso2.carbon.identity.template.mgt.endpoint/src/main/java/org/wso2/carbon/identity/template/mgt/endpoint/impl/TemplatesApiServiceImpl.java:58]: @GANGANI we are changing the backward compatibility in this deprecating effort. Any particular reasons for this? Can't we only add a depreciate tag?

File: components/template-mgt/org.wso2.carbon.identity.template.mgt.endpoint/src/main/java/org/wso2/carbon/identity/template/mgt/endpoint/impl/TemplatesApiServiceImpl.java (hunk 9 above)
Comment (GANGANI) [reply components/template-mgt/org.wso2.carbon.identity.template.mgt.endpoint/src/main/java/org/wso2/carbon/identity/template/mgt/endpoint/impl/TemplatesApiServiceImpl.java:51]: I didn't change the access modifier, I set it to one line.

File: components/template-mgt/org.wso2.carbon.identity.template.mgt.endpoint/src/main/java/org/wso2/carbon/identity/template/mgt/endpoint/impl/TemplatesApiServiceImpl.java (hunk 10 above)
Comment (GANGANI) [reply components/template-mgt/org.wso2.carbon.identity.template.mgt.endpoint/src/main/java/org/wso2/carbon/identity/template/mgt/endpoint/impl/TemplatesApiServiceImpl.java:58]: As per the discussion with @emswbandara and the team, we decide to return the forbidden error.

**Pull request merged**